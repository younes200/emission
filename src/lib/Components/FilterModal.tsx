import { Box, Button, CloseIcon, Flex, Sans } from "@artsy/palette"
import React from "react"
import { LayoutAnimation, Modal as RNModal, TouchableWithoutFeedback, ViewProperties } from "react-native"
import styled from "styled-components/native"

interface ModalProps extends ViewProperties {
  visible?: boolean
  closeModal?: () => void
}

interface State {
  isComponentMounted: boolean
}

const ModalBackgroundView = styled.View`
  background-color: #00000099;
  flex: 1;
  flex-direction: column;
`

const ModalInnerView = styled.View<{ visible: boolean }>`
  flex-direction: column;
  width: 100%;
  background-color: white;
  height: ${({ visible }) => (visible ? "auto" : "0")};
  padding: ${({ visible }) => (visible ? "20px" : "0")};
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
`

export class FilterModal extends React.Component<ModalProps, State> {
  constructor(props) {
    super(props)

    this.state = { isComponentMounted: false }
  }

  componentDidMount() {
    setTimeout(() => {
      LayoutAnimation.configureNext(LayoutAnimation.Presets.easeInEaseOut)
      this.setState({ isComponentMounted: true })
    }, 100)
  }

  closeModal() {
    if (this.props.closeModal) {
      this.props.closeModal()
    }
  }

  render() {
    /*
    https://app.zeplin.io/project/5aabc5e0786bbb29b6e3dc7f/screen/5dba03df09bf3f5547162ac9

    Next steps:

    - Add a Navigator
      - Configure left button to be x close
      - Configure right button for "clear all"
    - Define a TypeScript interface to represent possible filter states (take a look at the Pick TypeScript generic)
    - Implement the rest of the UI
    - Add callbacks for when filter has been cleared / reset / applied

    Let's start with a single filter (sort, or medium maybe) and get it working end-to-end in Collections
    Look at Reactions' ArtworkFilter implementation, which does something similar ^
    */
    return (
      <RNModal animationType="fade" transparent={true} visible={this.props.visible}>
        <TouchableWithoutFeedback>
          <ModalBackgroundView>
            <TouchableWithoutFeedback onPress={null}>
              <>
                <Flex onTouchStart={() => this.closeModal()} style={{ flexGrow: 1 }} />
                <ModalInnerView visible={this.state.isComponentMounted}>
                  <Flex flexDirection="row" justifyContent="space-between">
                    <Flex alignItems="flex-end" mt={0.5} mb={2}>
                      <Box onTouchEnd={() => this.closeModal()}>
                        <CloseIcon fill="black100" />
                      </Box>
                    </Flex>
                    <Sans weight="medium" size="4">
                      Filter
                    </Sans>
                    <Sans size="4">Clear all</Sans>
                  </Flex>
                  {this.props.children}
                  <Button
                    onPress={() => {
                      this.closeModal()
                    }}
                    block
                    width={100}
                    variant="secondaryOutline"
                  >
                    Ok
                  </Button>
                </ModalInnerView>
              </>
            </TouchableWithoutFeedback>
          </ModalBackgroundView>
        </TouchableWithoutFeedback>
      </RNModal>
    )
  }
}
