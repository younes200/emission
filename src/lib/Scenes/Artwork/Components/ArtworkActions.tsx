import {
  BellFillIcon,
  BellIcon,
  Box,
  color,
  EyeOpenedIcon,
  Flex,
  HeartFillIcon,
  HeartIcon,
  Sans,
  ShareIcon,
} from "@artsy/palette"
import { ArtworkActions_artwork } from "__generated__/ArtworkActions_artwork.graphql"
import { ArtworkActionsSaveMutation } from "__generated__/ArtworkActionsSaveMutation.graphql"
import Events from "lib/NativeModules/Events"
import { Schema, track } from "lib/utils/track"
import { take } from "lodash"
import React from "react"
import { NativeModules, Share, TouchableWithoutFeedback, View } from "react-native"
import { commitMutation, createFragmentContainer, graphql, RelayProp } from "react-relay"
import styled from "styled-components/native"

const Constants = NativeModules.ARCocoaConstantsModule
const ApiModule = NativeModules.ARTemporaryAPIModule

interface ArtworkActionsProps {
  artwork: ArtworkActions_artwork
  relay?: RelayProp
}

export const shareContent = (title, href, artists: ArtworkActions_artwork["artists"]) => {
  let computedTitle: string | null
  if (artists && artists.length) {
    const names = take(artists, 3).map(artist => artist.name)
    computedTitle = `${title} by ${names.join(", ")} on Artsy`
  } else if (title) {
    computedTitle = `${title} on Artsy`
  }
  return {
    title: computedTitle,
    message: computedTitle,
    url: `https://artsy.net${href}`,
  }
}

@track()
export class ArtworkActions extends React.Component<ArtworkActionsProps> {
  @track((props: ArtworkActionsProps) => {
    return {
      action_name: props.artwork.is_saved ? Schema.ActionNames.ArtworkUnsave : Schema.ActionNames.ArtworkSave,
      action_type: Schema.ActionTypes.Success,
      context_module: Schema.ContextModules.ArtworkActions,
    }
  })
  handleArtworkSave() {
    const { artwork, relay } = this.props
    commitMutation<ArtworkActionsSaveMutation>(relay.environment, {
      mutation: graphql`
        mutation ArtworkActionsSaveMutation($input: SaveArtworkInput!) {
          saveArtwork(input: $input) {
            artwork {
              id
              is_saved: isSaved
            }
          }
        }
      `,
      variables: { input: { artworkID: artwork.internalID, remove: artwork.is_saved } },
      optimisticResponse: { saveArtwork: { artwork: { id: artwork.id, is_saved: !artwork.is_saved } } },
      onCompleted: () => Events.userHadMeaningfulInteraction(),
    })
  }

  @track(() => ({
    action_name: Schema.ActionNames.Share,
    action_type: Schema.ActionTypes.Tap,
    context_module: Schema.ContextModules.ArtworkActions,
  }))
  async handleArtworkShare() {
    const { title, href, artists } = this.props.artwork
    try {
      await Share.share(shareContent(title, href, artists))
    } catch (error) {
      console.error("ArtworkActions.tsx", error)
    }
  }

  @track(() => ({
    action_name: Schema.ActionNames.ViewInRoom,
    action_type: Schema.ActionTypes.Tap,
    context_module: Schema.ContextModules.ArtworkActions,
  }))
  openViewInRoom() {
    const {
      artwork: { image, id, slug, heightCm, widthCm },
    } = this.props
    const heightIn = heightCm / 2.54
    const widthIn = widthCm / 2.54

    ApiModule.presentAugmentedRealityVIR(image.url, widthIn, heightIn, slug, id)
  }

  render() {
    const {
      artwork: { is_saved, is_hangable, sale },
    } = this.props

    const isOpenSale = sale && sale.isAuction && !sale.isClosed

    return (
      <View>
        <Flex flexDirection="row">
          {isOpenSale ? (
            <TouchableWithoutFeedback onPress={() => this.handleArtworkSave()}>
              <UtilButton pr={3}>
                <Box mr={0.5}>{is_saved ? <BellFillIcon fill="purple100" /> : <BellIcon />}</Box>
                <Sans size="3" color={is_saved ? color("purple100") : color("black100")}>
                  Watch lot
                </Sans>
              </UtilButton>
            </TouchableWithoutFeedback>
          ) : (
            <TouchableWithoutFeedback onPress={() => this.handleArtworkSave()}>
              <UtilButton pr={3}>
                <Box mr={0.5}>{is_saved ? <HeartFillIcon fill="purple100" /> : <HeartIcon />}</Box>
                <Sans size="3" color={is_saved ? color("purple100") : color("black100")}>
                  {is_saved ? "Saved" : "Save"}
                </Sans>
              </UtilButton>
            </TouchableWithoutFeedback>
          )}

          {Constants.AREnabled && is_hangable && (
            <TouchableWithoutFeedback onPress={() => this.openViewInRoom()}>
              <UtilButton pr={3}>
                <Box mr={0.5}>
                  <EyeOpenedIcon />
                </Box>
                <Sans size="3">View in Room</Sans>
              </UtilButton>
            </TouchableWithoutFeedback>
          )}
          <TouchableWithoutFeedback onPress={() => this.handleArtworkShare()}>
            <UtilButton>
              <Box mr={0.5}>
                <ShareIcon />
              </Box>
              <Sans size="3">Share</Sans>
            </UtilButton>
          </TouchableWithoutFeedback>
        </Flex>
      </View>
    )
  }
}

const UtilButton = styled(Flex)`
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
`

export const ArtworkActionsFragmentContainer = createFragmentContainer(ArtworkActions, {
  artwork: graphql`
    fragment ArtworkActions_artwork on Artwork {
      id
      internalID
      slug
      title
      href
      is_saved: isSaved
      is_hangable: isHangable
      artists {
        name
      }
      image {
        url
      }
      sale {
        isAuction
        isClosed
      }
      widthCm
      heightCm
    }
  `,
})
