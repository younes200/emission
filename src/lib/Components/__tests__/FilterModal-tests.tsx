import { Theme } from "@artsy/palette"
import React from "react"

import "react-native"
import * as renderer from "react-test-renderer"

import { FilterModal } from "lib/Components/FilterModal"

it("looks like expected", () => {
  let props

  props = {
    visible: true,
  }

  const tree = renderer
    .create(
      <Theme>
        <FilterModal {...props} />
      </Theme>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
