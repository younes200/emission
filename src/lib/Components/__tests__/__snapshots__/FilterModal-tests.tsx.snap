// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`looks like expected 1`] = `
<Modal
  animationType="fade"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#00000099",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ]
    }
  >
    <View
      onTouchStart={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
          },
          Object {
            "flexGrow": 1,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "backgroundColor": "white",
            "borderTopLeftRadius": 10,
            "borderTopRightRadius": 10,
            "flexDirection": "column",
            "height": 0,
            "paddingBottom": 0,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 0,
            "width": "100%",
          },
        ]
      }
      visible={false}
    >
      <View
        flexDirection="row"
        justifyContent="space-between"
        style={
          Array [
            Object {
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
          ]
        }
      >
        <View
          alignItems="flex-end"
          mb={2}
          mt={0.5}
          style={
            Array [
              Object {
                "alignItems": "flex-end",
                "display": "flex",
                "marginBottom": 20,
                "marginTop": 5,
              },
            ]
          }
        >
          <View
            onTouchStart={[Function]}
            style={
              Array [
                Object {},
              ]
            }
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight="18px"
              bbWidth="18px"
              className="Iconios__Icon-z4d8jr-0 cjvLsm"
              fill="black100"
              focusable={false}
              height="18px"
              meetOrSlice={0}
              minX={0}
              minY={0}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  undefined,
                  Object {
                    "opacity": 1,
                  },
                  Object {
                    "flex": 0,
                    "height": 18,
                    "width": 18,
                  },
                ]
              }
              vbHeight={18}
              vbWidth={18}
              width="18px"
            >
              <RNSVGGroup
                fill={
                  Array [
                    0,
                    4278190080,
                  ]
                }
                fillOpacity={1}
                fillRule={1}
                font={Object {}}
                matrix={
                  Array [
                    1,
                    0,
                    0,
                    1,
                    0,
                    0,
                  ]
                }
                opacity={1}
                originX={0}
                originY={0}
                propList={
                  Array [
                    "fill",
                  ]
                }
                rotation={0}
                scaleX={1}
                scaleY={1}
                skewX={0}
                skewY={0}
                stroke={null}
                strokeDasharray={null}
                strokeDashoffset={null}
                strokeLinecap={0}
                strokeLinejoin={0}
                strokeMiterlimit={4}
                strokeOpacity={1}
                strokeWidth={1}
                vectorEffect={0}
                x={0}
                y={0}
              >
                <RNSVGPath
                  d="M9.88 9l4.56 4.56-.88.88L9 9.88l-4.56 4.56-.88-.88L8.12 9 3.56 4.44l.88-.88L9 8.12l4.56-4.56.88.88z"
                  fill={
                    Array [
                      0,
                      4278190080,
                    ]
                  }
                  fillOpacity={1}
                  fillRule={0}
                  matrix={
                    Array [
                      1,
                      0,
                      0,
                      1,
                      0,
                      0,
                    ]
                  }
                  opacity={1}
                  originX={0}
                  originY={0}
                  propList={
                    Array [
                      "fill",
                      "fillRule",
                    ]
                  }
                  rotation={0}
                  scaleX={1}
                  scaleY={1}
                  skewX={0}
                  skewY={0}
                  stroke={null}
                  strokeDasharray={null}
                  strokeDashoffset={null}
                  strokeLinecap={0}
                  strokeLinejoin={0}
                  strokeMiterlimit={4}
                  strokeOpacity={1}
                  strokeWidth={1}
                  vectorEffect={0}
                  x={0}
                  y={0}
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
        </View>
        <Text
          fontFamily="Unica77LL-Medium"
          fontSize="16px"
          lineHeight="26px"
          style={
            Array [
              Object {
                "fontFamily": "Unica77LL-Medium",
                "fontSize": 16,
                "lineHeight": 26,
              },
              Object {},
            ]
          }
        >
          Filter
        </Text>
        <Text
          fontFamily="Unica77LL-Regular"
          fontSize="16px"
          lineHeight="26px"
          style={
            Array [
              Object {
                "fontFamily": "Unica77LL-Regular",
                "fontSize": 16,
                "lineHeight": 26,
              },
              Object {},
            ]
          }
        >
          Clear all
        </Text>
      </View>
      <View
        accessible={true}
        flexDirection="row"
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "display": "flex",
              "flexDirection": "row",
            },
          ]
        }
      >
        <View
          block={true}
          onPress={[Function]}
          px={2}
          size="medium"
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 3,
                "borderWidth": 1,
                "justifyContent": "center",
                "paddingLeft": 20,
                "paddingRight": 20,
                "position": "relative",
                "width": "100%",
              },
              Object {
                "backgroundColor": "rgba(255, 255, 255, 1)",
                "borderColor": "rgba(229, 229, 229, 1)",
                "color": "rgba(0, 0, 0, 1)",
                "height": 41,
                "opacity": 1,
              },
            ]
          }
          theme={
            Object {
              "borders": Array [
                "1px solid",
                "2px solid",
              ],
              "breakpoints": Array [
                768,
                900,
                1024,
                1192,
              ],
              "colors": Object {
                "black10": "#E5E5E5",
                "black100": "#000",
                "black30": "#C2C2C2",
                "black5": "#F8F8F8",
                "black60": "#666",
                "black80": "#333",
                "green100": "#0EDA83",
                "purple100": "#6E1EFF",
                "purple30": "#D3BBFF",
                "purple5": "#F8F3FF",
                "red100": "#F7625A",
                "white100": "#FFF",
                "yellow10": "#FDF7E8",
                "yellow100": "#F1AF1B",
                "yellow30": "#FAE7BA",
              },
              "fontFamily": Object {
                "display": Object {
                  "regular": "AvantGardeGothicITC",
                },
                "sans": Object {
                  "italic": "Unica77LL-Italic",
                  "medium": "Unica77LL-Medium",
                  "mediumItalic": "Unica77LL-MediumItalic",
                  "regular": "Unica77LL-Regular",
                },
                "serif": Object {
                  "italic": "ReactNativeAGaramondPro-Italic",
                  "regular": "ReactNativeAGaramondPro-Regular",
                  "semibold": "ReactNativeAGaramondPro-Semibold",
                },
              },
              "grid": Object {
                "breakpoints": Object {
                  "lg": 1024,
                  "md": 900,
                  "sm": 768,
                  "xl": 1192,
                  "xs": 767,
                },
                "col": Object {
                  "padding": 0,
                },
                "container": Object {
                  "padding": 0,
                },
                "row": Object {
                  "padding": 0,
                },
              },
              "mediaQueries": Object {
                "hover": "not all and (pointer: coarse), not all and (-moz-touch-enabled: 1)",
                "lg": "(min-width: 1024px) and (max-width: 1191px)",
                "md": "(min-width: 900px) and (max-width: 1023px)",
                "sm": "(min-width: 768px) and (max-width: 899px)",
                "xl": "(min-width: 1192px)",
                "xs": "(max-width: 767px)",
              },
              "space": Object {
                "0.3": 3,
                "0.5": 5,
                "1": 10,
                "12": 120,
                "18": 180,
                "2": 20,
                "3": 30,
                "4": 40,
                "6": 60,
                "9": 90,
              },
              "typeSizes": Object {
                "display": Object {
                  "2": Object {
                    "fontSize": 10,
                    "lineHeight": 12,
                  },
                  "3t": Object {
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                  "4t": Object {
                    "fontSize": 14,
                    "lineHeight": 18,
                  },
                  "5t": Object {
                    "fontSize": 16,
                    "lineHeight": 20,
                  },
                  "6": Object {
                    "fontSize": 18,
                    "lineHeight": 22,
                  },
                  "8": Object {
                    "fontSize": 22,
                    "lineHeight": 24,
                  },
                },
                "sans": Object {
                  "0": Object {
                    "fontSize": 8,
                    "lineHeight": 8,
                  },
                  "1": Object {
                    "fontSize": 10,
                    "lineHeight": 14,
                  },
                  "10": Object {
                    "fontSize": 42,
                    "lineHeight": 50,
                  },
                  "12": Object {
                    "fontSize": 60,
                    "lineHeight": 66,
                  },
                  "14": Object {
                    "fontSize": 80,
                    "lineHeight": 84,
                  },
                  "16": Object {
                    "fontSize": 100,
                    "lineHeight": 104,
                  },
                  "2": Object {
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                  "3": Object {
                    "fontSize": 14,
                    "lineHeight": 24,
                  },
                  "3t": Object {
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  "4": Object {
                    "fontSize": 16,
                    "lineHeight": 26,
                  },
                  "4t": Object {
                    "fontSize": 16,
                    "lineHeight": 22,
                  },
                  "5": Object {
                    "fontSize": 18,
                    "lineHeight": 30,
                  },
                  "5t": Object {
                    "fontSize": 18,
                    "lineHeight": 26,
                  },
                  "6": Object {
                    "fontSize": 22,
                    "lineHeight": 30,
                  },
                  "8": Object {
                    "fontSize": 28,
                    "lineHeight": 36,
                  },
                },
                "serif": Object {
                  "1": Object {
                    "fontSize": 12,
                    "lineHeight": 16,
                  },
                  "10": Object {
                    "fontSize": 44,
                    "lineHeight": 50,
                  },
                  "12": Object {
                    "fontSize": 60,
                    "lineHeight": 70,
                  },
                  "2": Object {
                    "fontSize": 14,
                    "lineHeight": 18,
                  },
                  "3": Object {
                    "fontSize": 16,
                    "lineHeight": 24,
                  },
                  "3t": Object {
                    "fontSize": 16,
                    "lineHeight": 20,
                  },
                  "4": Object {
                    "fontSize": 18,
                    "lineHeight": 26,
                  },
                  "4t": Object {
                    "fontSize": 18,
                    "lineHeight": 22,
                  },
                  "5": Object {
                    "fontSize": 22,
                    "lineHeight": 32,
                  },
                  "5t": Object {
                    "fontSize": 22,
                    "lineHeight": 28,
                  },
                  "6": Object {
                    "fontSize": 26,
                    "lineHeight": 32,
                  },
                  "8": Object {
                    "fontSize": 32,
                    "lineHeight": 38,
                  },
                },
              },
            }
          }
          variant="secondaryOutline"
          width={100}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "display": "flex",
                  "flexDirection": "row",
                  "height": "100%",
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": "100%",
                },
              ]
            }
          >
            <Text
              color="#000"
              fontFamily="Unica77LL-Medium"
              fontSize="14px"
              lineHeight="20px"
              style={
                Array [
                  Object {
                    "color": "#000",
                    "fontFamily": "Unica77LL-Medium",
                    "fontSize": 14,
                    "lineHeight": 20,
                  },
                  Object {},
                ]
              }
            >
              Ok
            </Text>
          </View>
          <Text
            fontFamily="Unica77LL-Medium"
            fontSize="14px"
            lineHeight="20px"
            role="presentation"
            style={
              Array [
                Object {
                  "fontFamily": "Unica77LL-Medium",
                  "fontSize": 14,
                  "lineHeight": 20,
                },
                Object {
                  "opacity": 0,
                },
              ]
            }
          >
            Ok
          </Text>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
